### PR TITLE
Fix the split learning example

### PR DESCRIPTION
**Pull request by cuiboyuan**
Update the split learning example code to accommodate changes from Plato

## Description

With the new updates in the Plato codes, the original implementation of split learning is no longer working, so this PR just updates the example code to fix that.

The following files are modified:

`split_learning_algorithm.py`: update such that the model weights that are not trained by clients will not be affected by the updates from clients. Also update the two training methods for clients to make use of the GPU device if exists. Lastly, store the data used to extract features to ensure that the order of input training data for clients is the same as the order of gradients received from the server

`split_learning_server.py`: some clean-ups to avoid duplicate codes.

`split_learning_client.py`: some refactoring to accommodate new changes from Plato

`split_learning_trainer.py`: update logging codes and remove the use of the sampler to ensure that the order of the gradients to be sent to the clients is the same as the order of the features received from the clients

`split_learning_MNIST_lenet5.yml`: lower the learning rate as it seems that the model weights on the clients tend to become `NaN` during the training. I'm not sure whether this is due to the current implementation or if it is a problem that is expected for split learning. Since for split learning, when the server is calculating the gradients for the clients, they are calculated based on the assumption that the part of the model that is on the client is not changing. But when clients are backpropagating with these gradients, the model weights are changing, and after the first few batches of gradients, the gradients from the server might be leading the model weights on the client to the wrong direction, preventing the model from converging.

## How has this been tested?

Run `python examples/split_learning/split_learning.py -c examples/split_learning/split_learning_MNIST_lenet5.yml` without errors

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue) Fixes #
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by cuiboyuan** [comment]
Hi @baochunli ,

I've taken another look into the split learning config files, and it seems like the splitNN is more likely to work well in the scenario where each client only trains on a small number of data. When the training set size on the client is small, the client and the server will communicate more frequently, making the training of splitNN more similar to the training of a local model without cut layers.

Additionally, I think this discovery is also in alignment with what is demonstrated in the original paper. Although the author does not mention about the data size on each client, he/she only conducted the experiment in the scenario of a large number of clients (i.e. 100 and 500 clients). Even though it is not explicitly mentioned, I think the data is distributed in an IID manner (since it is probably the most common way to partition the data among clients), which will make the number of data on each client to be small. 

I ran the split learning with 500 clients and 100 data example on each for 200 rounds, and below is the change of accuracy:
split-learning-acc

The model was able to reach an accuracy of around 93% at the end. Below is the config yaml file used to run the test:

```
clients:
    # Type
    type: split_learning

    # The total number of clients
    total_clients: 500

    # The number of clients selected in each round
    per_round: 1

    # Should the clients compute test accuracy locally?
    do_test: false

server:
    type: split_learning

    address: 127.0.0.1
    port: 8000

data:
    # The training and testing dataset
    datasource: MNIST

    # Number of samples in each partition
    partition_size: 100

    # Fixed random seed
    random_seed: 1

    # IID, biased, or sharded?
    sampler: iid

trainer:
    # The type of the trainer
    type: split_learning

    # The maximum number of training rounds
    rounds: 200

    # The maximum number of clients running concurrently
    max_concurrency: 3

    # The target accuracy
    target_accuracy: 0.95

    # The machine learning model
    model_name: lenet5

    # Number of epoches for local training in each communication round
    epochs: 1
    batch_size: 32
    optimizer: SGD
    
algorithm:
    # Aggregation algorithm
    type: split_learning

    # Split learning flag
    split_learning: true

parameters:
    model:
        num_classes: 10
        cut_layer: relu4

    optimizer:
        lr: 0.01
        momentum: 0.9
        weight_decay: 0.0

results: round, accuracy
```

Hope this helps.

Thanks,

**Comment by baochunli** [comment]
> Hi @baochunli ,
> 
> I've taken another look into the split learning config files, and it seems like the splitNN is more likely to work well in the scenario where each client only trains on a small number of data. When the training set size on the client is small, the client and the server will communicate more frequently, making the training of splitNN more similar to the training of a local model without cut layers.
> 
> Additionally, I think this discovery is also in alignment with what is demonstrated in the original paper. Although the author does not mention about the data size on each client, he/she only conducted the experiment in the scenario of a large number of clients (i.e. 100 and 500 clients). Even though it is not explicitly mentioned, I think the data is distributed in an IID manner (since it is probably the most common way to partition the data among clients), which will make the number of data on each client to be small.
> 
> I ran the split learning with 500 clients and 100 data example on each for 200 rounds, and below is the change of accuracy: split-learning-acc
> 
> The model was able to reach an accuracy of around 93% at the end. Below is the config yaml file used to run the test:
> 
> ```
> clients:
>     # Type
>     type: split_learning
> 
>     # The total number of clients
>     total_clients: 500
> 
>     # The number of clients selected in each round
>     per_round: 1
> 
>     # Should the clients compute test accuracy locally?
>     do_test: false
> 
> server:
>     type: split_learning
> 
>     address: 127.0.0.1
>     port: 8000
> 
> data:
>     # The training and testing dataset
>     datasource: MNIST
> 
>     # Number of samples in each partition
>     partition_size: 100
> 
>     # Fixed random seed
>     random_seed: 1
> 
>     # IID, biased, or sharded?
>     sampler: iid
> 
> trainer:
>     # The type of the trainer
>     type: split_learning
> 
>     # The maximum number of training rounds
>     rounds: 200
> 
>     # The maximum number of clients running concurrently
>     max_concurrency: 3
> 
>     # The target accuracy
>     target_accuracy: 0.95
> 
>     # The machine learning model
>     model_name: lenet5
> 
>     # Number of epoches for local training in each communication round
>     epochs: 1
>     batch_size: 32
>     optimizer: SGD
>     
> algorithm:
>     # Aggregation algorithm
>     type: split_learning
> 
>     # Split learning flag
>     split_learning: true
> 
> parameters:
>     model:
>         num_classes: 10
>         cut_layer: relu4
> 
>     optimizer:
>         lr: 0.01
>         momentum: 0.9
>         weight_decay: 0.0
> 
> results: round, accuracy
> ```
> 
> Hope this helps.
> 
> Thanks,

Sorry for the late response; I finally got a bit of time to get back to this PR. What you said is very insightful indeed. The code works with the latest version of Plato, but it hasn't been adapted to the Plato's new client and server public API (see documentation for details). We are refactoring the code to make it compatible, and to change several places where the quality of the codebase is not good.

**Comment by cuiboyuan** [comment]
I see; that sounds great. Thanks for the reply!
